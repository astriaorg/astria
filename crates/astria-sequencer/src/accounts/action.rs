use anyhow::{
    ensure,
    Context,
    Result,
};
use astria_core::{
    primitive::v1::Address,
    protocol::transaction::v1alpha1::action::TransferAction,
};
use tracing::instrument;

use crate::{
    accounts::state_ext::{
        StateReadExt,
        StateWriteExt,
    },
    bridge::state_ext::StateReadExt as _,
    state_ext::{
        StateReadExt as _,
        StateWriteExt as _,
    },
    transaction::action_handler::ActionHandler,
};

pub(crate) async fn transfer_check_stateful<S: StateReadExt + 'static>(
    action: &TransferAction,
    state: &S,
    from: Address,
) -> Result<()> {
    ensure!(
        state
            .is_allowed_fee_asset(action.fee_asset_id)
            .await
            .context("failed to check allowed fee assets in state")?,
        "invalid fee asset",
    );

    let fee = state
        .get_transfer_base_fee()
        .await
        .context("failed to get transfer base fee")?;
    let transfer_asset_id = action.asset_id;

    let from_fee_balance = state
        .get_account_balance(from, action.fee_asset_id)
        .await
        .context("failed getting `from` account balance for fee payment")?;

    // if fee asset is same as transfer asset, ensure accounts has enough funds
    // to cover both the fee and the amount transferred
    if action.fee_asset_id == transfer_asset_id {
        let payment_amount = action
            .amount
            .checked_add(fee)
            .context("transfer amount plus fee overflowed")?;

        ensure!(
            from_fee_balance >= payment_amount,
            "insufficient funds for transfer and fee payment"
        );
    } else {
        // otherwise, check the fee asset account has enough to cover the fees,
        // and the transfer asset account has enough to cover the transfer
        ensure!(
            from_fee_balance >= fee,
            "insufficient funds for fee payment"
        );

        let from_transfer_balance = state
            .get_account_balance(from, transfer_asset_id)
            .await
            .context("failed to get account balance in transfer check")?;
        ensure!(
            from_transfer_balance >= action.amount,
            "insufficient funds for transfer"
        );
    }

    Ok(())
}

#[async_trait::async_trait]
impl ActionHandler for TransferAction {
    async fn check_stateful<S: StateReadExt + 'static>(
        &self,
        state: &S,
        from: Address,
    ) -> Result<()> {
        // ensure the recipient is not a bridge account,
        // as the explicit `BridgeLockAction` should be used to transfer to a bridge account.
        ensure!(
            state
                .get_bridge_account_rollup_id(&self.to)
                .await
                .context("failed to get bridge account rollup ID from state")?
                .is_none(),
            "cannot send transfer to bridge account",
        );

        transfer_check_stateful(self, state, from)
            .await
            .context("stateful transfer check failed")
    }

    #[instrument(
        skip_all,
        fields(
            to = self.to.to_string(),
            amount = self.amount,
        )
    )]
    async fn execute<S: StateWriteExt>(&self, state: &mut S, from: Address) -> Result<()> {
        let fee = state
            .get_transfer_base_fee()
            .await
            .context("failed to get transfer base fee")?;
        state
            .get_and_increase_block_fees(self.fee_asset_id, fee)
            .await
            .context("failed to add to block fees")?;

        let transfer_asset_id = self.asset_id;

        // if fee payment asset is same asset as transfer asset, deduct fee
        // from same balance as asset transferred
        if transfer_asset_id == self.fee_asset_id {
            // check_stateful should have already checked this arithmetic
            let payment_amount = self
                .amount
                .checked_add(fee)
                .expect("transfer amount plus fee should not overflow");

            state
                .decrease_balance(from, transfer_asset_id, payment_amount)
                .await
                .context("failed decreasing `from` account balance")?;
            state
                .increase_balance(self.to, transfer_asset_id, self.amount)
                .await
                .context("failed increasing `to` account balance")?;
        } else {
            // otherwise, just transfer the transfer asset and deduct fee from fee asset balance
            // later
            state
                .decrease_balance(from, transfer_asset_id, self.amount)
                .await
                .context("failed decreasing `from` account balance")?;
            state
                .increase_balance(self.to, transfer_asset_id, self.amount)
                .await
                .context("failed increasing `to` account balance")?;

            // deduct fee from fee asset balance
            state
                .decrease_balance(from, self.fee_asset_id, fee)
                .await
                .context("failed decreasing `from` account balance for fee payment")?;
        }

        Ok(())
    }
}
