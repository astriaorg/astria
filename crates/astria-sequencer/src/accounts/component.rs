use std::sync::Arc;

use anyhow::{
    Context,
    Result,
};
use tendermint::abci::request::{
    BeginBlock,
    EndBlock,
};
use tracing::instrument;

use super::state_ext::StateWriteExt;
use crate::{
    asset::get_native_asset,
    component::Component,
    genesis::GenesisState,
};

#[derive(Default)]
pub(crate) struct AccountsComponent;

#[async_trait::async_trait]
impl Component for AccountsComponent {
    type AppState = GenesisState;

    #[instrument(name = "AccountsComponent:init_chain", skip(state))]
    async fn init_chain<S: StateWriteExt>(mut state: S, app_state: &Self::AppState) -> Result<()> {
        let native_asset = get_native_asset();
        for account in &app_state.accounts {
            state
                .put_account_balance(account.address, native_asset.id(), account.balance)
                .context("failed writing account balance to state")?;
        }
        state
            .put_ibc_sudo_address(app_state.ibc_sudo_address)
            .context("failed to set IBC sudo key")?;
        Ok(())
    }

    #[instrument(name = "AccountsComponent:begin_block", skip(_state))]
    async fn begin_block<S: StateWriteExt + 'static>(
        _state: &mut Arc<S>,
        _begin_block: &BeginBlock,
    ) -> Result<()> {
        Ok(())
    }

    #[instrument(name = "AccountsComponent:end_block", skip(_state))]
    async fn end_block<S: StateWriteExt + 'static>(
        _state: &mut Arc<S>,
        _end_block: &EndBlock,
    ) -> Result<()> {
        Ok(())
    }
}
