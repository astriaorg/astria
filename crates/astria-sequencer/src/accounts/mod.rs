pub(crate) mod action;
pub(crate) mod component;
pub(crate) mod ics20_transfer;
pub(crate) mod query;
pub(crate) mod state_ext;
