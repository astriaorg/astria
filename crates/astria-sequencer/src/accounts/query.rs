use anyhow::Context as _;
use penumbra_storage::{
    Snapshot,
    Storage,
};
use proto::native::sequencer::v1alpha1::Address;
use tendermint::{
    abci::{
        request,
        response,
    },
    block::Height,
};

use crate::{
    accounts::state_ext::StateReadExt as _,
    service::AbciCode,
    state_ext::StateReadExt as _,
};

pub(crate) async fn balance_request(
    storage: Storage,
    request: request::Query,
    params: Vec<(String, String)>,
) -> response::Query {
    use proto::{
        native::sequencer::v1alpha1::BalanceResponse,
        Message as _,
    };
    let (address, snapshot, height) = match preprocess_request(&storage, &request, &params).await {
        Ok(tup) => tup,
        Err(err_rsp) => return err_rsp,
    };
    let balance = match snapshot.get_account_balance(address).await {
        Ok(balance) => balance,
        Err(err) => {
            return response::Query {
                code: AbciCode::INVALID_PARAMETER.into(),
                info: format!("{}", AbciCode::INVALID_PARAMETER),
                log: format!("failed getting balance for provided address: {err:?}"),
                height,
                ..response::Query::default()
            };
        }
    };
    let payload = BalanceResponse {
        height: height.value(),
        balance,
    }
    .into_raw()
    .encode_to_vec()
    .into();
    response::Query {
        code: 0.into(),
        key: request.path.clone().into_bytes().into(),
        value: payload,
        height,
        ..response::Query::default()
    }
}

pub(crate) async fn nonce_request(
    storage: Storage,
    request: request::Query,
    params: Vec<(String, String)>,
) -> response::Query {
    use proto::{
        native::sequencer::v1alpha1::NonceResponse,
        Message as _,
    };
    let (address, snapshot, height) = match preprocess_request(&storage, &request, &params).await {
        Ok(tup) => tup,
        Err(err_rsp) => return err_rsp,
    };
    let nonce = match snapshot.get_account_nonce(address).await {
        Ok(nonce) => nonce,
        Err(err) => {
            return response::Query {
                code: 2.into(),
                info: "failed getting nonce for provided address".into(),
                log: format!("{err:?}"),
                height,
                ..response::Query::default()
            };
        }
    };
    let payload = NonceResponse {
        height: height.value(),
        nonce,
    }
    .into_raw()
    .encode_to_vec()
    .into();
    response::Query {
        code: AbciCode::OK.into(),
        info: format!("{}", AbciCode::OK),
        key: request.path.clone().into_bytes().into(),
        value: payload,
        height,
        ..response::Query::default()
    }
}

async fn get_snapshot_and_height(
    storage: &Storage,
    height: Height,
) -> anyhow::Result<(Snapshot, Height)> {
    let snapshot = match height.value() {
        0 => storage.latest_snapshot(),
        other => {
            let version = storage
                .latest_snapshot()
                .get_storage_version_by_height(other)
                .await
                .context("failed to get storage version from height")?;
            storage
                .snapshot(version)
                .context("failed to get storage at version")?
        }
    };
    let height: Height = snapshot
        .get_block_height()
        .await
        .context("failed to get block height from snapshot")?
        .try_into()
        .context("internal u64 block height does not fit into tendermint i64 `Height`")?;
    Ok((snapshot, height))
}

async fn preprocess_request(
    storage: &Storage,
    request: &request::Query,
    params: &[(String, String)],
) -> anyhow::Result<(Address, Snapshot, Height), response::Query> {
    let Some(address) = params
        .iter()
        .find_map(|(k, v)| (k == "account").then_some(v))
    else {
        return Err(response::Query {
            code: AbciCode::INVALID_PARAMETER.into(),
            info: AbciCode::INVALID_PARAMETER.to_string(),
            log: "path did not contain path parameter".into(),
            ..response::Query::default()
        });
    };
    let address = hex::decode(address)
        .context("failed decoding hex encoded bytes")
        .and_then(|addr| {
            Address::try_from_slice(&addr).context("failed constructing address from bytes")
        })
        .map_err(|err| response::Query {
            code: AbciCode::INVALID_PARAMETER.into(),
            info: AbciCode::INVALID_PARAMETER.to_string(),
            log: format!("address could not be constructed from provided parameter: {err:?}"),
            ..response::Query::default()
        })?;
    let (snapshot, height) = match get_snapshot_and_height(storage, request.height).await {
        Ok(tup) => tup,
        Err(err) => {
            return Err(response::Query {
                code: AbciCode::INTERNAL_ERROR.into(),
                info: AbciCode::INTERNAL_ERROR.to_string(),
                log: format!("failed to query internal storage for snapshot and height: {err:?}"),
                ..response::Query::default()
            });
        }
    };
    Ok((address, snapshot, height))
}
