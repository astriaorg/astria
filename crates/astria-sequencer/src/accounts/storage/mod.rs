pub(super) mod keys;
mod values;

pub(crate) use values::Value;
pub(super) use values::{
    Balance,
    Nonce,
};
