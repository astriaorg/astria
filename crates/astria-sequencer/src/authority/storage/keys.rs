use crate::accounts::AddressBytes;

pub(in crate::authority) const SUDO: &str = "authority/sudo";
pub(in crate::authority) const VALIDATOR_PREFIX: &str = "authority/validator/";
pub(in crate::authority) const VALIDATOR_COUNT: &str = "authority/validator_count";
pub(in crate::authority) const VALIDATOR_UPDATES: &str = "authority/validator_updates";

pub(in crate::authority) const PRE_ASPEN_VALIDATOR_SET: &str = "authority/validator_set"; // Deprecated post Aspen upgrade

pub(in crate::authority) fn validator<TAddress: AddressBytes>(address: &TAddress) -> String {
    format!(
        "{}/{}",
        VALIDATOR_PREFIX,
        hex::encode(address.address_bytes())
    )
}

#[cfg(test)]
mod tests {
    use astria_core::{
        crypto::ADDRESS_LENGTH,
        primitive::v1::Address,
    };

    use super::*;

    const COMPONENT_PREFIX: &str = "authority/";

    fn snapshot_validator() -> String {
        validator(
            &Address::builder()
                .array([1; ADDRESS_LENGTH])
                .prefix("astria")
                .try_build()
                .unwrap(),
        )
    }

    #[test]
    fn keys_should_not_change() {
        insta::assert_snapshot!("sudo_address_key", SUDO);
        insta::assert_snapshot!("validator_key", snapshot_validator());
        insta::assert_snapshot!("validator_count_key", VALIDATOR_COUNT);

        insta::assert_snapshot!("validator_set_key", PRE_ASPEN_VALIDATOR_SET);
        insta::assert_snapshot!("validator_updates_key", VALIDATOR_UPDATES);
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(SUDO.starts_with(COMPONENT_PREFIX));
        assert!(snapshot_validator().starts_with(COMPONENT_PREFIX));
        assert!(VALIDATOR_COUNT.starts_with(COMPONENT_PREFIX));

        assert!(PRE_ASPEN_VALIDATOR_SET.starts_with(COMPONENT_PREFIX));
        assert!(VALIDATOR_UPDATES.starts_with(COMPONENT_PREFIX));
    }
}
