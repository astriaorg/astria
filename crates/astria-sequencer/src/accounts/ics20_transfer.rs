//! This module implements the ICS20 transfer handler, which handles
//! incoming packets.
//!
//! It contains an [`Ics20Transfer`] struct which implements the Penumbra
//! [`AppHandler`] trait, which is passed through the Penumbra IBC implementation
//! during transaction checks and execution. The IBC implementation calls into
//! the ICS20 transfer handler during the IBC transaction lifecycle.
//!
//! [`AppHandler`] consists of two traits: [`AppHandlerCheck`] and [`AppHandlerExecute`].
//! [`AppHandlerCheck`] is used for stateless and stateful checks, while
//! [`AppHandlerExecute`] is used for execution.

use anyhow::{
    Context as _,
    Result,
};
use astria_core::sequencer::v1alpha1::{
    asset::IbcAsset,
    Address,
};
use cnidarium::{
    StateRead,
    StateWrite,
};
use ibc_types::{
    core::channel::{
        channel,
        msgs::{
            MsgAcknowledgement,
            MsgChannelCloseConfirm,
            MsgChannelCloseInit,
            MsgChannelOpenAck,
            MsgChannelOpenConfirm,
            MsgChannelOpenInit,
            MsgChannelOpenTry,
            MsgRecvPacket,
            MsgTimeout,
        },
        ChannelId,
        PortId,
    },
    transfer::acknowledgement::TokenTransferAcknowledgement,
};
use penumbra_ibc::component::app_handler::{
    AppHandler,
    AppHandlerCheck,
    AppHandlerExecute,
};
use penumbra_proto::penumbra::core::component::ibc::v1alpha1::FungibleTokenPacketData;

use super::state_ext::{
    StateReadExt as _,
    StateWriteExt,
};
use crate::asset::state_ext::{
    StateReadExt as _,
    StateWriteExt as _,
};

/// The ICS20 transfer handler.
///
/// See [here](https://github.com/cosmos/ibc/blob/main/spec/app/ics-020-fungible-token-transfer/README.md)
/// for the specification which this is based on.
#[derive(Clone)]
pub(crate) struct Ics20Transfer;

#[async_trait::async_trait]
impl AppHandlerCheck for Ics20Transfer {
    async fn chan_open_init_check<S: StateRead>(_: S, msg: &MsgChannelOpenInit) -> Result<()> {
        if msg.ordering != channel::Order::Unordered {
            anyhow::bail!("channel order must be unordered for Ics20 transfer");
        }

        if msg.version_proposal.as_str() != "ics20-1" {
            anyhow::bail!("channel version must be ics20-1 for Ics20 transfer");
        }

        Ok(())
    }

    async fn chan_open_try_check<S: StateRead>(_: S, msg: &MsgChannelOpenTry) -> Result<()> {
        if msg.ordering != channel::Order::Unordered {
            anyhow::bail!("channel order must be unordered for Ics20 transfer");
        }

        if msg.version_supported_on_a.as_str() != "ics20-1" {
            anyhow::bail!("counterparty version must be ics20-1 for Ics20 transfer");
        }

        Ok(())
    }

    async fn chan_open_ack_check<S: StateRead>(_: S, msg: &MsgChannelOpenAck) -> Result<()> {
        if msg.version_on_b.as_str() != "ics20-1" {
            anyhow::bail!("counterparty version must be ics20-1 for Ics20 transfer");
        }

        Ok(())
    }

    async fn chan_open_confirm_check<S: StateRead>(_: S, _: &MsgChannelOpenConfirm) -> Result<()> {
        // accept channel confirmations, port has already been validated, version has already been
        // validated
        Ok(())
    }

    async fn chan_close_init_check<S: StateRead>(_: S, _: &MsgChannelCloseInit) -> Result<()> {
        anyhow::bail!("ics20 always aborts on chan_close_init");
    }

    async fn chan_close_confirm_check<S: StateRead>(
        _: S,
        _: &MsgChannelCloseConfirm,
    ) -> Result<()> {
        // no action needed
        Ok(())
    }

    async fn recv_packet_check<S: StateRead>(_: S, _: &MsgRecvPacket) -> Result<()> {
        // checks performed in `execute`
        Ok(())
    }

    async fn timeout_packet_check<S: StateRead>(state: S, msg: &MsgTimeout) -> Result<()> {
        refund_tokens_check(
            state,
            msg.packet.data.as_slice(),
            &msg.packet.port_on_a,
            &msg.packet.chan_on_a,
        )
        .await
    }

    async fn acknowledge_packet_check<S: StateRead>(
        state: S,
        msg: &MsgAcknowledgement,
    ) -> Result<()> {
        // see https://github.com/cosmos/ibc-go/blob/3f5b2b6632e0fa37056e5805b289a9307870ac9a/modules/core/04-channel/types/acknowledgement.go
        // and https://github.com/cosmos/ibc-go/blob/3f5b2b6632e0fa37056e5805b289a9307870ac9a/proto/ibc/core/channel/v1/channel.proto#L155
        // for formatting
        let ack: TokenTransferAcknowledgement =
            serde_json::from_slice(msg.acknowledgement.as_slice())?;
        if ack.is_successful() {
            return Ok(());
        }

        refund_tokens_check(
            state,
            msg.packet.data.as_slice(),
            &msg.packet.port_on_a,
            &msg.packet.chan_on_a,
        )
        .await
    }
}

async fn refund_tokens_check<S: StateRead>(
    state: S,
    data: &[u8],
    source_port: &PortId,
    source_channel: &ChannelId,
) -> Result<()> {
    use prost::Message as _;

    let packet_data = FungibleTokenPacketData::decode(data)
        .context("failed to decode fungible token packet data json")?;
    let mut asset = packet_data
        .denom
        .parse::<IbcAsset>()
        .context("failed parsing `denom` field packet data as IbcAsset")?;

    // if the asset is prefixed with `ibc`, the rest of the denomination string is the asset ID,
    // so we need to look up the full trace from storage.
    // see https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-001-coin-source-tracing.md#decision
    if asset.prefix_is("ibc") {
        asset = state
            .get_ibc_asset(asset.id())
            .await
            .context("failed to get denom trace from asset id")?;
    }

    if is_source(source_port, source_channel, &asset, true) {
        // sender of packet (us) was the source chain
        //
        // check if escrow account has enough balance to refund user
        let balance = state
            .get_ibc_channel_balance(source_channel, asset.id())
            .await
            .context("failed to get channel balance in refund_tokens_check")?;

        let packet_amount: u128 = packet_data
            .amount
            .parse()
            .context("failed to parse packet amount as u128")?;
        if balance < packet_amount {
            anyhow::bail!("insufficient balance to refund tokens to sender");
        }
    }

    Ok(())
}

fn is_source(
    source_port: &PortId,
    source_channel: &ChannelId,
    asset: &IbcAsset,
    is_refund: bool,
) -> bool {
    let prefix = format!("{source_port}/{source_channel}/");
    if is_refund {
        !asset.prefix_is(&prefix)
    } else {
        asset.prefix_is(&prefix)
    }
}

#[async_trait::async_trait]
impl AppHandlerExecute for Ics20Transfer {
    async fn chan_open_init_execute<S: StateWrite>(_: S, _: &MsgChannelOpenInit) {}

    async fn chan_open_try_execute<S: StateWrite>(_: S, _: &MsgChannelOpenTry) {}

    async fn chan_open_ack_execute<S: StateWrite>(_: S, _: &MsgChannelOpenAck) {}

    async fn chan_open_confirm_execute<S: StateWrite>(_: S, _: &MsgChannelOpenConfirm) {}

    async fn chan_close_confirm_execute<S: StateWrite>(_: S, _: &MsgChannelCloseConfirm) {}

    async fn chan_close_init_execute<S: StateWrite>(_: S, _: &MsgChannelCloseInit) {}

    async fn recv_packet_execute<S: StateWrite>(mut state: S, msg: &MsgRecvPacket) {
        use penumbra_ibc::component::packet::WriteAcknowledgement as _;

        let ack = match execute_ics20_transfer(
            &mut state,
            &msg.packet.data,
            &msg.packet.port_on_a,
            &msg.packet.chan_on_a,
            &msg.packet.port_on_b,
            &msg.packet.chan_on_b,
            false,
        )
        .await
        {
            Ok(()) => TokenTransferAcknowledgement::success(),
            Err(e) => TokenTransferAcknowledgement::Error(e.to_string()),
        };

        let ack_bytes: Vec<u8> = ack.into();

        if let Err(e) = state.write_acknowledgement(&msg.packet, &ack_bytes).await {
            let error: &dyn std::error::Error = e.as_ref();
            tracing::error!(error, "failed to write acknowledgement");
        }
    }

    async fn timeout_packet_execute<S: StateWrite>(mut state: S, msg: &MsgTimeout) {
        // we put source and dest as chain_a (the source) as we're refunding tokens,
        // and the destination chain of the refund is the source.
        if let Err(e) = execute_ics20_transfer(
            &mut state,
            &msg.packet.data,
            &msg.packet.port_on_a,
            &msg.packet.chan_on_a,
            &msg.packet.port_on_a,
            &msg.packet.chan_on_a,
            true,
        )
        .await
        {
            let error: &dyn std::error::Error = e.as_ref();
            tracing::error!(
                error,
                "failed to refund tokens during timeout_packet_execute",
            );
        };
    }

    async fn acknowledge_packet_execute<S: StateWrite>(mut state: S, msg: &MsgAcknowledgement) {
        let ack: TokenTransferAcknowledgement = serde_json::from_slice(
            msg.acknowledgement.as_slice(),
        )
        .expect("valid acknowledgement, should have been checked in acknowledge_packet_check");
        if ack.is_successful() {
            return;
        }

        // we put source and dest as chain_a (the source) as we're refunding tokens,
        // and the destination chain of the refund is the source.
        if let Err(e) = execute_ics20_transfer(
            &mut state,
            &msg.packet.data,
            &msg.packet.port_on_a,
            &msg.packet.chan_on_a,
            &msg.packet.port_on_a,
            &msg.packet.chan_on_a,
            true,
        )
        .await
        {
            let error: &dyn std::error::Error = e.as_ref();
            tracing::error!(
                error,
                "failed to refund tokens during acknowledge_packet_execute",
            );
        }
    }
}

#[async_trait::async_trait]
impl AppHandler for Ics20Transfer {}

async fn execute_ics20_transfer<S: StateWriteExt>(
    state: &mut S,
    data: &[u8],
    source_port: &PortId,
    source_channel: &ChannelId,
    dest_port: &PortId,
    dest_channel: &ChannelId,
    is_refund: bool,
) -> Result<()> {
    let packet_data: FungibleTokenPacketData =
        serde_json::from_slice(data).context("failed to decode FungibleTokenPacketData")?;
    let packet_amount: u128 = packet_data
        .amount
        .parse()
        .context("failed to parse packet data amount to u128")?;
    let recipient = Address::try_from_slice(
        &hex::decode(packet_data.receiver).context("failed to decode receiver as hex string")?,
    )
    .context("invalid receiver address")?;
    let mut asset = packet_data
        .denom
        .parse::<IbcAsset>()
        .context("failed parsing `denom` field packet data as IbcAsset")?;

    // if the asset is prefixed with `ibc`, the rest of the denomination string is the asset ID,
    // so we need to look up the full trace from storage.
    // see https://github.com/cosmos/ibc-go/blob/main/docs/architecture/adr-001-coin-source-tracing.md#decision
    if asset.prefix_is("ibc") {
        asset = state
            .get_ibc_asset(asset.id())
            .await
            .context("failed to get denom trace from asset id")?;
    }

    if is_source(source_port, source_channel, &asset, is_refund) {
        // sender of packet (us) was the source chain
        // subtract balance from escrow account and transfer to user

        let escrow_balance = state
            .get_ibc_channel_balance(source_channel, asset.id())
            .await
            .context("failed to get IBC channel balance in execute_ics20_transfer")?;

        let user_balance = state.get_account_balance(recipient, asset.id()).await?;
        state
            .put_ibc_channel_balance(
                source_channel,
                asset.id(),
                escrow_balance
                    .checked_sub(packet_amount)
                    .ok_or(anyhow::anyhow!(
                        "insufficient balance in escrow account to transfer tokens"
                    ))?,
            )
            .context("failed to update escrow account balance in execute_ics20_transfer")?;
        state
            .put_account_balance(
                recipient,
                asset.id(),
                user_balance
                    .checked_add(packet_amount)
                    .ok_or(anyhow::anyhow!("overflow when adding to user balance"))?,
            )
            .context("failed to update user account balance in execute_ics20_transfer")?;
    } else {
        let prefixed_denomination = if is_refund {
            // we're refunding a token we issued and tried to bridge, but failed
            packet_data.denom
        } else {
            // we're receiving a token from another chain
            // create a token with additional prefix and mint it to the recipient
            format!("{dest_port}/{dest_channel}/{}", packet_data.denom)
        };

        let asset: IbcAsset = prefixed_denomination
            .parse()
            .context("failed to parse prefixed denomination as IbcAsset")?;

        // register denomination in global ID -> denom map if it's not already there
        if !state
            .has_ibc_asset(asset.id())
            .await
            .context("failed to check if ibc asset exists in state")?
        {
            state
                .put_ibc_asset(asset.id(), &asset)
                .context("failed to put IBC asset in storage")?;
        }

        let user_balance = state
            .get_account_balance(recipient, asset.id())
            .await
            .context("failed to get user account balance in execute_ics20_transfer")?;
        state
            .put_account_balance(
                recipient,
                asset.id(),
                user_balance
                    .checked_add(packet_amount)
                    .ok_or(anyhow::anyhow!("overflow when adding to user balance"))?,
            )
            .context("failed to update user account balance in execute_ics20_transfer")?;
    }

    Ok(())
}
