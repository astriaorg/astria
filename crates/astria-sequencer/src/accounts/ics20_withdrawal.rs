use anyhow::{
    ensure,
    Context as _,
    Result,
};
use ibc_types::core::channel::{
    ChannelId,
    PortId,
};
use penumbra_ibc::component::packet::{
    IBCPacket,
    SendPacketRead as _,
    SendPacketWrite as _,
    Unchecked,
};
use proto::native::sequencer::v1alpha1::{
    asset,
    asset::IbcAsset,
    Address,
    Ics20Withdrawal,
};
use tracing::instrument;

use crate::{
    accounts::state_ext::{
        StateReadExt,
        StateWriteExt,
    },
    transaction::action_handler::ActionHandler,
};

fn withdrawal_to_unchecked_ibc_packet(withdrawal: &Ics20Withdrawal) -> IBCPacket<Unchecked> {
    let packet_data = withdrawal.to_fungible_token_packet_data();
    let serialized_packet_data =
        serde_json::to_vec(&packet_data).expect("can serialize FungibleTokenPacketData as JSON");

    IBCPacket::new(
        PortId::transfer(),
        withdrawal.source_channel().clone(),
        *withdrawal.timeout_height(),
        withdrawal.timeout_time(),
        serialized_packet_data,
    )
}

#[async_trait::async_trait]
impl ActionHandler for Ics20Withdrawal {
    #[instrument(skip(self))]
    async fn check_stateless(&self) -> Result<()> {
        ensure!(self.timeout_time() != 0, "timeout time must be non-zero",);

        // NOTE (from penumbra): we could validate the destination chain address as bech32 to
        // prevent mistyped addresses, but this would preclude sending to chains that don't
        // use bech32 addresses.
        Ok(())
    }

    #[instrument(skip(self, state))]
    async fn check_stateful<S: StateReadExt + 'static>(
        &self,
        state: &S,
        from: Address,
        _fee_asset_id: asset::Id,
    ) -> Result<()> {
        let packet: IBCPacket<Unchecked> = withdrawal_to_unchecked_ibc_packet(self);
        state
            .send_packet_check(packet)
            .await
            .context("packet failed send check")?;

        let from_transfer_balance = state
            .get_account_balance(from, self.denom().id())
            .await
            .context("failed getting `from` account balance for transfer")?;
        ensure!(
            from_transfer_balance >= self.amount(),
            "insufficient funds for transfer"
        );

        Ok(())
    }

    #[instrument(skip(self, state))]
    async fn execute<S: StateWriteExt>(
        &self,
        state: &mut S,
        from: Address,
        _fee_asset_id: asset::Id,
    ) -> Result<()> {
        let checked_packet = withdrawal_to_unchecked_ibc_packet(self).assume_checked();

        let from_transfer_balance = state
            .get_account_balance(from, self.denom().id())
            .await
            .context("failed getting `from` account balance for transfer")?;

        state
            .put_account_balance(
                from,
                self.denom().id(),
                from_transfer_balance
                    .checked_sub(self.amount())
                    .context("insufficient funds for transfer")?,
            )
            .context("failed to update sender balance")?;

        // if we're the source, move tokens to the escrow account,
        // otherwise the tokens are just burned
        if is_source(
            checked_packet.source_port(),
            checked_packet.source_channel(),
            self.denom(),
        ) {
            let channel_balance = state
                .get_ibc_channel_balance(self.source_channel(), self.denom().id())
                .await
                .context("failed to get channel balance")?;

            state
                .put_ibc_channel_balance(
                    self.source_channel(),
                    self.denom().id(),
                    channel_balance
                        .checked_add(self.amount())
                        .context("overflow when adding to channel balance")?,
                )
                .context("failed to update channel balance")?;
        }

        state.send_packet_execute(checked_packet).await;
        Ok(())
    }
}

fn is_source(source_port: &PortId, source_channel: &ChannelId, asset: &IbcAsset) -> bool {
    let prefix = format!("{source_port}/{source_channel}/");
    !asset.prefix_is(&prefix)
}
