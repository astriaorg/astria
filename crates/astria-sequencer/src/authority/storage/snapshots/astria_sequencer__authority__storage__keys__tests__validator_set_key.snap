---
source: crates/astria-sequencer/src/authority/storage/keys.rs
expression: VALIDATOR_SET
---
authority/validator_set
